servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpServicePayloadDiscardWatchdogTest.java

```diff
@@ -0,0 +1,104 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.netty;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.http.api.BlockingHttpClient;
+import io.servicetalk.http.api.HttpResponse;
+import io.servicetalk.http.api.HttpServerContext;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.transport.api.HostAndPort;
+
+import org.junit.jupiter.api.Test;
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+final class HttpServicePayloadDiscardWatchdogTest {
+
+    @Test
+    void cleansPayloadBodyIfDiscardedInFilter() throws Exception {
+        final AtomicBoolean payloadSubscribed = new AtomicBoolean(false);
+        try (HttpServerContext serverContext = HttpServers
+                .forPort(0)
+                .appendServiceFilter(service -> new StreamingHttpServiceFilter(service) {
+                    @Override
+                    public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                                final StreamingHttpRequest request,
+                                                                final StreamingHttpResponseFactory responseFactory) {
+                        return delegate()
+                                .handle(ctx, request, responseFactory)
+                                .beforeOnSuccess(res -> res.transformPayloadBody(payload -> Publisher.empty()));
+                    }
+                })
+                .listenStreamingAndAwait((ctx, request, responseFactory) -> {
+                    final Publisher<Buffer> buffer = Publisher
+                            .from(ctx.executionContext().bufferAllocator().fromUtf8("Hello, World!"))
+                            .beforeOnSubscribe(subscription -> payloadSubscribed.set(true));
+                    return Single.succeeded(responseFactory.ok().payloadBody(buffer));
+                })) {
+            try (BlockingHttpClient client = HttpClients.forSingleAddress(
+                    HostAndPort.of((InetSocketAddress) serverContext.listenAddress())).buildBlocking()) {
+                HttpResponse response = client.request(client.get("/"));
+                assertEquals(0, response.payloadBody().readableBytes());
+            }
+
+            assertTrue(payloadSubscribed.get());
+        }
+    }
+
+    @Test
+    void cleansPayloadBodyIfErrorInFilter() throws Exception {
```

Comment:
Consider adding a test that uses `map` to drop the original response and create a new one